packages/wareki/src/index.ts

```diff
@@ -0,0 +1,130 @@
+const { WAREKI_START_YEARS, reg, selectGengo } = (() => {
+  const REIWA = '令和' as const
+  const HEISEI = '平成' as const
+  const SHOWA = '昭和' as const
+  const TAISHO = '大正' as const
+  const MEIJI = '明治' as const
+  type Geongo = typeof REIWA | typeof HEISEI | typeof SHOWA | typeof TAISHO | typeof MEIJI
+
+  const SEPARATOR = '[:\\/\\-\\.\\s．年月日]'
+  const WAREKI_BOUNDARYS = [
+    [REIWA, 2019, 4, 30, HEISEI],
+    [HEISEI, 1989, 1, 7, SHOWA],
+    [SHOWA, 1926, 12, 24, TAISHO],
+    [TAISHO, 1912, 7, 29, MEIJI],
+  ] as const
+  const YEARS = {
+    t: 1912,
+    T: 1912,
+    [TAISHO]: 1912,
+    s: 1926,
+    S: 1926,
+    [SHOWA]: 1926,
+    h: 1989,
+    H: 1989,
+    [HEISEI]: 1989,
+    r: 2019,
+    R: 2019,
+    [REIWA]: 2019,
+    m: 1868,
+    M: 1868,
+    [MEIJI]: 1868,
+  } as const
```

Comment:
[IMO] こちらの定数は「各元号の始まりの年」と「各元号を表す記号」をまとめて表現しているために定数の意味がわかりにくくなっている気がして、例えば以下のようそれぞれの意味を分けて表現すれば定数の意味は明確になるかもと思いました。

```suggestion
const WAREKI_START_YEARS = {
  [MEIJI]: 1868,
  [TAISHO]: 1912,
  [SHOWA]: 1926,
  [HEISEI]: 1989,
  [REIWA]: 2019,
} as const
const WAREKI_SIGNS = {
  m: MEIJI,
  M: MEIJI,
  [MEIJI]: MEIJI,
  t: TAISHO,
  T: TAISHO,
  [TAISHO]: TAISHO,
  s: SHOWA,
  S: SHOWA,
  [SHOWA]: SHOWA,
  h: HEISEI,
  H: HEISEI,
  [HEISEI]: HEISEI,
  r: REIWA,
  R: REIWA,
  [REIWA]: REIWA,
} as const
```

Reply:
もともとの思想は
- ページ初期表示時に連発される可能性がある（データとしては西暦で持っているため）
だったため、`m -> 1868 ` の変換などはなるべく高速にしたいなぁ、という感じでこのようにしたはず... (微々たるものですが)
↑のデータ構造にすればデータ構造自体は読みやすくなりそうですが、使う側が `WAREKI_START_YEARS [WAREKI_SIGNS[year]]` みたいな構造になって微妙な予感がします...

やるとしたら以下の様にSTART_YEARを持って置くと良いかも？
```
const TAISYO_START_YEAR = 1912
  const YEARS = {
    t: TAISYO_START_YEAR,
    T: TAISYO_START_YEAR,
    [TAISHO]: TAISYO_START_YEAR,
    ...
  } as const
```

Reply:
最終的な判断はお任せしますので、一応見解だけ述べておくと

オブジェクトが一つから二つになることによるパフォーマンスの悪化はほぼ誤差レベルだと思うので、それで可読性が買えるなら安いと考えることもできると思います。

> 使う側が `WAREKI_START_YEARS[WAREKI_SIGNS[year]]` みたいな構造になって微妙

こちらに関しては、1行で書くのであれば確かにそうなのですが

```ts
const wareki = WAREKI_SIGNS[sign]
const startingYear = WAREKI_START_YEAR[wareki]
```

と書き下せば特に問題なく意味も明確にできるので、使い方次第かなと思いました。